07_StagingAndMultipleQueues/main.cpp

```diff
@@ -73,7 +84,414 @@ class StagingAndMultipleQueuesApp final : public examples::BasicMultiQueueApplic
 			return retval;
 		}
 
+		template<typename... Args>
+		void logFailAndTerminate(const char* msg, Args&&... args)
+		{
+			m_logger->log(msg, ILogger::ELL_ERROR, std::forward<Args>(args)...);
+			std::exit(-1);
+		}
+
+	private:
+		std::array<core::smart_refctd_ptr<IGPUImage>, IMAGE_CNT> images;
+
+		smart_refctd_ptr<IGPUBuffer> histogramBuffer = nullptr;
+		nbl::video::IDeviceMemoryAllocator::SMemoryOffset histogramBufferAllocation = {};
+
+		std::mutex imagesReadyMutex;
+		std::condition_variable imagesReady;
+		bool imagesReadyFlag = false;
+
+		std::mutex histogramBufferReadyMutex;
+		std::condition_variable histogramBufferReady;
+		bool histogramBufferReadyFlag = false;
```

Comment:
here's the idea, the Upload, Compute, Download **need to overlap**

So create `FramesInFlight` (can be 4 when you'll use timeline semaphores, but for now have IMG_CNT numer) commandbuffers and `notify_one` after you submit each one of them

Also change the Flags to **atomic_uint32_t**
You can actually get rid of the condition_variable and use wait-on address introduced with C++20 for atomic counters

Then the producer thread increments and tells us how many items are ready and the consumer thread spinds using wait on address until the value increments realtive to the "processed count"